File: analysis/revision/scripts/capstone.py
import pandas as pd, numpy as np
df=pd.read_csv('./data/Tree_ID_info.csv'); df['site']=np.where(df['plot']=='BGS','BGS','EMS')
sm={'rm':'A.rubrum','bg':'N.sylvatica','ro':'Q.rubra','hem':'T.canadensis'}; df['sp']=df['species'].map(sm)
ert=pd.read_csv('./data/ERT_application_results.csv'); m=df.merge(ert,on='tree').reset_index(drop=True)
mets=['mean','median','sd','cv','gini','entropy','cma','radialgradient']
Z=m[mets].copy()
for c in mets: Z[c]=m.groupby('sp')[c].transform(lambda x:(x-x.mean())/x.std(ddof=1))
Zc=Z.values-Z.values.mean(0); U,S,Vt=np.linalg.svd(Zc,full_matrices=False); pc=(U*S)[:,0]
if np.corrcoef(pc,m['mean'])[0,1]>0: pc=-pc
m['pc1']=pc; m['struct']=m['percent_damaged']>1
mb=(m.site=='BGS').values&(~m['struct'].values); me=(m.site=='EMS').values&(~m['struct'].values)
nb=(~m['struct'].values&(m.site=='BGS').values).sum(); ne=(~m['struct'].values&(m.site=='EMS').values).sum()
# species-standardized wetness (higher=wetter)
xdev=-m.groupby('sp')['mean'].transform(lambda x:(x-x.median())/x.std(ddof=1)).values
# pooled PC1
Zp=(m[mets]-m[mets].mean())/m[mets].std(ddof=1); Zp=Zp.values-Zp.values.mean(0)
Up,Sp,Vtp=np.linalg.svd(Zp,full_matrices=False); pcp=(Up*Sp)[:,0]
if np.corrcoef(pcp,m['mean'])[0,1]>0: pcp=-pcp

sd=pc.std(); sdx=xdev.std(); sdp=pcp.std()
# each axis: (values array where higher=more anomalous/wet, list of thresholds spanning -1..+1 SD around its center)
axes={
 'PC1 (within-sp)': (pc, [pc.mean()+k*sd for k in (-1,-0.5,-0.25,0,0.25,0.5,1)]),
 'PC1 (pooled)':    (pcp,[pcp.mean()+k*sdp for k in (-1,-0.5,-0.25,0,0.25,0.5,1)]),
 'species-median resistivity': (xdev,[0+k*sdx for k in (-1,-0.5,-0.25,0,0.25,0.5,1)]),
 'absolute resistivity': (-m['mean'].values, sorted(-np.quantile(m['mean'].values,[0.35,0.42,0.5,0.58,0.65]))),
 'CMA': (m['cma'].values, [0.25,0.30,0.33,0.36,0.40]),
}
print("SITE DIRECTION ROBUSTNESS: incipient among structurally-sound trees, BGS vs EMS")
print(f"{'axis':28s}{'#thr':>5}{'BGS>EMS holds':>15}{'BGS% range':>16}{'EMS% range':>16}")
total=0; held=0
for name,(v,ths) in axes.items():
    bs=[]; es=[]; ok=0
    for t in ths:
        anom=v>t
        b=(anom&mb).sum()/nb*100; e=(anom&me).sum()/ne*100
        bs.append(b); es.append(e); ok+= (b>=e)
    total+=len(ths); held+=ok
    print(f"{name:28s}{len(ths):>5}{ok:>8}/{len(ths):<6}{min(bs):>6.0f}-{max(bs):<9.0f}{min(es):>6.0f}-{max(es):<9.0f}")
print(f"\nACROSS ALL {total} axis×threshold combinations: BGS>=EMS incipient in {held}/{total} = {held/total*100:.0f}%")

# structural prevalence stability by SoT threshold
print("\nStructural-loss prevalence by site, across SoT thresholds:")
for tau in [1,2,3,5]:
    s=m['percent_damaged']>tau
    b=s[m.site=='BGS'].mean()*100; e=s[m.site=='EMS'].mean()*100
    print(f"  SoT>{tau}%: BGS {b:.0f}% vs EMS {e:.0f}%  (EMS>BGS: {e>b})")

# species severity ordering stability (mean % damage among decayed, by species) — bootstrap-free descriptive
print("\nSpecies severity (mean % damage among decayed trees) — ordering:")
dd=m[m.percent_damaged>0]
print("  ", dd.groupby('sp')['percent_damaged'].mean().round(1).sort_values(ascending=False).to_dict())
print("  decayed n per species:", dd.groupby('sp').size().to_dict())

File: analysis/revision/scripts/capstone.R
# R port of capstone.py — site-direction robustness across axes and thresholds.
# Run from repo root: Rscript analysis/revision/scripts/capstone.R
source("analysis/revision/scripts/revision_common.R")

sd_pop <- function(x) sqrt(mean((x - mean(x))^2))  # numpy ndarray.std(), ddof=0

m <- load_merged()
bp <- build_pc1(m, "species")
pc <- bp$pc1
m$pc1 <- pc
m$struct <- m$percent_damaged > 1
mb <- m$site == "BGS" & !m$struct
me <- m$site == "EMS" & !m$struct
nb <- sum(mb); ne <- sum(me)
# species-standardized wetness (higher=wetter)
xdev <- xdev_species_median(m)
# pooled PC1
pcp <- build_pc1(m, "pooled")$pc1

sdv <- sd_pop(pc); sdx <- sd_pop(xdev); sdp <- sd_pop(pcp)
ks <- c(-1, -0.5, -0.25, 0, 0.25, 0.5, 1)
# each axis: (values where higher=more anomalous/wet, thresholds spanning -1..+1 SD)
axes <- list(
  "PC1 (within-sp)" = list(pc, mean(pc) + ks * sdv),
  "PC1 (pooled)"    = list(pcp, mean(pcp) + ks * sdp),
  "species-median resistivity" = list(xdev, 0 + ks * sdx),
  "absolute resistivity" = list(-m$mean,
    sort(-quantile(m$mean, c(0.35, 0.42, 0.5, 0.58, 0.65), type = 7, names = FALSE))),
  "CMA" = list(m$cma, c(0.25, 0.30, 0.33, 0.36, 0.40))
)

cat("SITE DIRECTION ROBUSTNESS: incipient among structurally-sound trees, BGS vs EMS\n")
cat(sprintf("%-28s%5s%15s%16s%16s\n", "axis", "#thr", "BGS>EMS holds", "BGS% range", "EMS% range"))
total <- 0; held <- 0
for (name in names(axes)) {
  vals <- axes[[name]][[1]]; ths <- axes[[name]][[2]]
  bs <- numeric(0); es <- numeric(0); ok <- 0
  for (t0 in ths) {
    anom <- vals > t0
    b <- sum(anom & mb) / nb * 100
    e <- sum(anom & me) / ne * 100
    bs <- c(bs, b); es <- c(es, e); ok <- ok + (b >= e)
  }
  total <- total + length(ths); held <- held + ok
  cat(sprintf("%-28s%5d%8d/%-6d%6.0f-%-9.0f%6.0f-%-9.0f\n",
              name, length(ths), ok, length(ths), min(bs), max(bs), min(es), max(es)))
}
cat(sprintf("\nACROSS ALL %d axis×threshold combinations: BGS>=EMS incipient in %d/%d = %.0f%%\n",
            total, held, total, held / total * 100))

# structural prevalence stability by SoT threshold
cat("\nStructural-loss prevalence by site, across SoT thresholds:\n")
for (tau in c(1, 2, 3, 5)) {
  s <- m$percent_damaged > tau
  b <- mean(s[m$site == "BGS"]) * 100
  e <- mean(s[m$site == "EMS"]) * 100
  cat(sprintf("  SoT>%d%%: BGS %.0f%% vs EMS %.0f%%  (EMS>BGS: %s)\n",
              tau, b, e, if (e > b) "True" else "False"))
}

# species severity ordering stability (mean % damage among decayed, by species)
cat("\nSpecies severity (mean % damage among decayed trees) — ordering:\n")
dd <- m[m$percent_damaged > 0, ]
sv <- sort(round(tapply(dd$percent_damaged, dd$sp, mean), 1), decreasing = TRUE)
cat("   {", paste(sprintf("'%s': %.1f", names(sv), sv), collapse = ", "), "}\n", sep = "")
cnt <- table(dd$sp)
cat("  decayed n per species: {", paste(sprintf("'%s': %d", names(cnt), cnt), collapse = ", "), "}\n", sep = "")
